src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs

```diff
@@ -0,0 +1,352 @@
+                {
+                    originalNodes.Add(startTag);
+                }
+                originalNodes.AddRange(tagChildren);
+                if (endTag != null)
+                {
+                    originalNodes.Add(endTag);
+                }
+
+                if (originalNodes.Count == 0)
+                {
+                    // Nothing to replace
+                    return parent;
+                }
+
+                // Replace nodes
+                var rewrittenElement = parent.ReplaceNodes(originalNodes, (original, rewritten) =>
+                {
+                    if (original == originalNodes[0] || rewritten == originalNodes[0])
+                    {
+                        // Put the new element in place of the first original node to be replaced.
+                        var body = new SyntaxList<RazorSyntaxNode>(tagChildren);
+                        var markupElement = SyntaxFactory.MarkupElement((MarkupTagBlockSyntax)startTag, body, endTag);
+                        return markupElement;
+                    }
+
+                    return null;
+                });
+
+                return rewrittenElement;
+            }
+
+            private bool TryRecoverStartTag(
+                SyntaxNode parent,
+                string tagName,
+                MarkupTagBlockSyntax endTag,
+                int endTagIndex,
+                out SyntaxNode rewritten,
+                out int startTagIndex)
+            {
+                rewritten = parent;
+                startTagIndex = -1;
+                var malformedTagCount = 0;
+
+                foreach (var tracker in _startTagTracker)
+                {
+                    if (tracker.TagName.Equals(tagName, StringComparison.OrdinalIgnoreCase))
+                    {
+                        break;
+                    }
+
+                    malformedTagCount++;
+                }
+
+                if (_startTagTracker.Count > malformedTagCount)
+                {
+                    parent = RewriteMalformedTags(parent, malformedTagCount);
+
+                    // One final rewrite, this is the rewrite that completes our target tag which is not malformed.
+                    var children = parent.ChildNodes();
+
+                    var startTagTracker = _startTagTracker.Pop();
+                    startTagIndex = startTagTracker.Index;
+                    var startTag = startTagTracker.TagBlock;
+                    
+                    // Get the nodes between the start and the end tag.
+                    var tagChildren = children.Skip(startTagIndex + 1).Take(endTagIndex - startTagIndex - 1).Cast<RazorSyntaxNode>();
+
+                    rewritten = RewriteNodeCore(parent, startTag, tagChildren, endTag);
+
+                    // We were able to recover
+                    return true;
+                }
+
+                // Could not recover tag. Aka we found an end tag without a corresponding start tag.
```

Comment:
Make sure you have a test for recovery when the cause is a malformed void element. We had a bug for this recently in Blazor.
```
<input ....></input>
```

Reply:
Added.